core/arch/riscv/mm/core_mmu.c

```diff
@@ -0,0 +1,639 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2022 NXP
+ */
+
+#include <assert.h>
+#include <bitstring.h>
+#include <config.h>
+#include <kernel/cache_helpers.h>
+#include <kernel/panic.h>
+#include <kernel/spinlock.h>
+#include <kernel/tee_l2cc_mutex.h>
+#include <kernel/tee_misc.h>
+#include <kernel/tlb_helpers.h>
+#include <mm/core_memprot.h>
+#include <mm/core_mmu.h>
+#include <platform_config.h>
+#include <riscv.h>
+#include <stdlib.h>
+#include <string.h>
+#include <trace.h>
+#include <util.h>
+
+#ifndef RV64
+#error implement
+#endif
+
+#ifndef DEBUG_XLAT_TABLE
+#define DEBUG_XLAT_TABLE 0
+#endif
+
+#if DEBUG_XLAT_TABLE
+#define debug_print(...) DMSG_RAW(__VA_ARGS__)
+#else
+#define debug_print(...) ((void)0)
+#endif
+
+bitstr_t g_asid __nex_bss;
+static unsigned int g_asid_spinlock __nex_bss = SPINLOCK_UNLOCK;
+
+typedef struct {
+	unsigned long entry;
+} pte_t;
+
+typedef struct {
+	pte_t entries[RISCV_PTES_PER_PT];
+} pgt_t;
+
+#define RISCV_MMU_PGT_SIZE	(RISCV_PTES_PER_PT * sizeof(pte_t))
+
+static pgt_t base_xlation_table
+	__aligned(RISCV_PGSIZE)
+	__section(".nozi.mmu.base_table");
+
+static pgt_t xlat_tables[RISCV_MMU_MAX_PGTS]
+	__aligned(RISCV_PGSIZE) __section(".nozi.mmu.l2");
+
+static pgt_t ta_tables[CFG_NUM_THREADS]
+	__aligned(RISCV_PGSIZE) __section(".nozi.mmu.l2");
+
+static int user_va_idx __nex_data = -1;
+
+struct mmu_partition {
+	pgt_t *base_table;
+	pgt_t *xlat_tables;
+	pgt_t *ta_tables;
+	unsigned int pgts_used;
```

Comment:
You have a few pgts above, which one does this concern?

Reply:
It concerns all tables. I see it as a pool of pages, when needed I allocate one from it. For now user va space and kernel va space are mapped within the same root page. It is fine for an initial implementation since user flag makes restriction.
But I would make 2 root page tables, one for user and one for kernel and set `satp` to the desired table when switching context between user/kernel. Still needs to better dig into `pgt_caches `first.